### DISCUSSION TITLE
O(1) Performance of editing member of observable array vs vanilla react

### DISCUSSION THREAD

**Discussion by 1mike12** [mobxjs/mobx#3569] [General] [answered]
So an interesting question popped up. I almost never messed around with immutable before. I went from default react states back in 2016, then recently came back and used mobx in 2021. The default way to do updates on an element of an array would be:

```js
class Example extends Component {
  constructor(){
     this.state = {array: [{name: "bob"}, {name: "john"}]}
  }
  changeName(index, name){
     const array = array.slice()
     array[i].name = name
     this.setState({array})
  }
}
```

We have an O(n) operation to make an entirely new array via `slice` in order to change the reference when React makes its shallow comparison pass. 

In mobx, we can modify arrays directly inside our actions. Is that modification O(1), or under the hood is there fundamentally also an O(n) operation?

**Comment by mweststrate** [comment]
O(1)

**Comment by kubk** [comment, marked as answer]
Also Redux has O(n) for number of subscribed components while Mobx has O(1): https://twitter.com/dan_abramov/status/719973322453348352?lang=en

**Reply by 1mike12** [reply]
Wow crazy, that's awesome.
